orderbook/src/price_estimate.rs

```diff
@@ -1,56 +1,155 @@
-use anyhow::{anyhow, ensure, Result};
-use ethcontract::H160;
-use model::TokenPair;
-use shared::uniswap_pool::PoolFetching;
-use std::iter::once;
+use anyhow::{anyhow, Result};
+use ethcontract::{H160, U256};
+use model::{order::OrderKind, TokenPair};
+use shared::{
+    uniswap_pool::{Pool, PoolFetching},
+    uniswap_solver::{
+        estimate_buy_amount, estimate_sell_amount, path_candidates, token_path_to_pair_path,
+    },
+};
+use std::{
+    cmp::Reverse,
+    collections::{HashMap, HashSet},
+};
+
+const MAX_HOPS: usize = 2;
 
 #[async_trait::async_trait]
 pub trait PriceEstimating: Send + Sync {
-    async fn estimate_price(&self, sell_token: H160, buy_token: H160) -> Result<f64>;
+    // Price is given in how much of sell_token needs to be sold for one buy_token.
+    async fn estimate_price(
+        &self,
+        sell_token: H160,
+        buy_token: H160,
+        amount: U256,
+        kind: OrderKind,
+    ) -> Result<f64>;
 }
 
 pub struct UniswapPriceEstimator {
     pool_fetcher: Box<dyn PoolFetching>,
+    base_tokens: HashSet<H160>,
 }
 
 impl UniswapPriceEstimator {
-    pub fn new(pool_fetcher: Box<dyn PoolFetching>) -> Self {
-        Self { pool_fetcher }
+    pub fn new(pool_fetcher: Box<dyn PoolFetching>, base_tokens: HashSet<H160>) -> Self {
+        Self {
+            pool_fetcher,
+            base_tokens,
+        }
     }
 }
 
 #[async_trait::async_trait]
 impl PriceEstimating for UniswapPriceEstimator {
-    // Estimates the price using the direct pool between sell and buy token. Price is given in
-    // how much of sell_token needs to be sold for one buy_token.
+    // Estimates the price using the direct pool between sell and buy token.
     // Returns an error if no pool exists between sell and buy token.
-    async fn estimate_price(&self, sell_token: H160, buy_token: H160) -> Result<f64> {
-        let pair = match TokenPair::new(sell_token, buy_token) {
-            Some(pair) => pair,
-            None => return Ok(1.0),
-        };
-        let pool = self
+    async fn estimate_price(
+        &self,
+        sell_token: H160,
+        buy_token: H160,
+        amount: U256,
+        kind: OrderKind,
+    ) -> Result<f64> {
```

Comment:
Could we include a doc comment on this method about which price is being estimated? I mean if its a sell order, then I assume the price is that of the buy token relative to the sell token (and vice-versa?). I just have the impression that the method would be self explanatory if, say, there were only a single token being passed in (but then it would be pegged to some reference token). Perhaps this obscurity could be mitigated by renaming the input parameters buy_token -> token and sell_token -> reference_token or maybe we could use base and quote.